# Good morning boss...

It was a tough year. We listened to all those scouts with their gut feelings, and didn't even make the playoffs.

I've compiled a list of the top 10 players for each position based off their value. I'm determining value by dividing a players OBP/Salary.

## Setting up the data

In [98]:
# Bring in the tools needed for the job
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re

from IPython.display import display, HTML

%matplotlib inline

# bring in the data
players = pd.read_csv('data/Master.csv', sep=',', parse_dates=['debut', 'finalGame'])
batting_stats = pd.read_csv('data/Batting.csv')
salaries = pd.read_csv('data/Salaries.csv')
positions = pd.read_csv('data/Appearances.csv')

# trim down some fields
salaries = salaries[['yearID', 'teamID', 'playerID', 'salary']]
players = players[['playerID', 'nameFirst', 'nameLast']]
salaries = salaries[['yearID', 'teamID', 'playerID', 'salary']]
batting_stats = batting_stats[['playerID', 'yearID', 'teamID', 'AB', 'H', 'BB', 'HBP', 'SF', 'R']]
positions = positions[['yearID', 'teamID', 'playerID', 'G_p', 'G_c', 'G_1b', 'G_2b', 'G_3b', 'G_ss', 'G_lf', 'G_cf', 'G_rf', 'G_of', 'G_dh']]

# merge data together
player_salaries = pd.merge(salaries, players)
sal_play_stats = pd.merge(player_salaries, batting_stats)

# OBP/Salary calculations
sps = sal_play_stats
sps['OBP'] = (sps['H']+sps['BB']+sps['HBP'])/(sps['AB']+sps['BB']+sps['HBP']+sps['SF'])
sps['value'] = sps['OBP']/(sps['salary']/100000)

# bring together player data with positions they played
sps = pd.merge(sps, positions)

In [112]:
def build_team(year, spend, ab):
    avail_pp = spend/9
    print('Money available per player: ', '${:,.2f}'.format(avail_pp))
    headers = ['nameFirst', 'nameLast', 'salary', 'R', 'AB', 'value']

    fb = sps[(sps.yearID == year) & (sps.G_1b > 1) & (sps.AB > ab) & (sps.salary < avail_pp)]
    fb = fb[headers]
    fb = fb.sort(['value'], ascending=False)[:10]
    fb['POS'] = '1B'
    count = len(fb)
    display(HTML('<h2>Top ' + str(count) + ' First Baseman</h2>'))
    display(fb)

    sb = sps[(sps.yearID == year) & (sps.G_2b > 1) & (sps.AB > ab) & (sps.salary < avail_pp)]
    sb = sb[headers]
    sb = sb.sort(['value'], ascending=False)[:10]
    sb['POS'] = '2B'
    count = len(sb)
    display(HTML('<h2>Top ' + str(count) + ' Second Baseman</h2>'))
    display(sb)


    ss = sps[(sps.yearID == year) & (sps.G_ss > 1) & (sps.AB > ab) & (sps.salary < avail_pp)]
    ss = ss[headers]
    ss = ss.sort(['value'], ascending=False)[:10]
    ss['POS'] = 'SS'
    count = len(ss)
    display(HTML('<h2>Top ' + str(count) + ' Short Stops</h2>'))
    display(ss)


    tb = sps[(sps.yearID == year) & (sps.G_3b > 1) & (sps.AB > ab) & (sps.salary < avail_pp)]
    tb = tb[headers]
    tb = tb.sort(['value'], ascending=False)[:10]
    tb['POS'] = '3B'
    display(HTML('<h2>Top ' + str(count) + ' Third Baseman</h2>'))
    display(tb)


    rf = sps[(sps.yearID == year) & (sps.G_rf > 1) & (sps.AB > ab) & (sps.salary < avail_pp)]
    rf = rf[headers]
    rf = rf.sort(['value'], ascending=False)[:10]
    rf['POS'] = 'RF'
    display(HTML('<h2>Top ' + str(count) + ' Right Fielders</h2>'))
    display(rf)


    cf = sps[(sps.yearID == year) & (sps.G_cf > 1) & (sps.AB > ab) & (sps.salary < avail_pp)]
    cf = cf[headers]
    cf = cf.sort(['value'], ascending=False)[:10]
    cf['POS'] = 'CF'
    display(HTML('<h2>Top ' + str(count) + ' Center Fielders</h2>'))
    display(cf)


    lf = sps[(sps.yearID == year) & (sps.G_lf > 1) & (sps.AB > ab) & (sps.salary < avail_pp)]
    lf = lf[headers]
    lf = lf.sort(['value'], ascending=False)[:10]
    lf['POS'] = 'LF'
    display(HTML('<h2>Top ' + str(count) + ' Left Fielders</h2>'))
    display(lf)


    c = sps[(sps.yearID == year) & (sps.G_c > 1) & (sps.AB > ab) & (sps.salary < avail_pp)]
    c = c[headers]
    c = c.sort(['value'], ascending=False)[:10]
    c['POS'] = 'C'
    display(HTML('<h2>Top ' + str(count) + ' Catchers</h2>'))
    display(c)

    p = sps[(sps.yearID == year) & (sps.G_p > 1) & (sps.AB > 20) & (sps.salary < avail_pp)]
    p = p[headers]
    p = p.sort(['value'], ascending=False)[:10]
    p['POS'] = 'P'
    display(HTML('<h2>Top ' + str(count) + ' Pitchers</h2>'))
    display(p)

In [116]:
build_team(2014, 5000000, 50)

Money available per player:  $555,555.56


nameFirst    nameLast  salary   R   AB     value POS
23570     Scott   Van Slyke  507500  32  212  0.076094  1B
23325    Lonnie  Chisenhall  510800  62  478  0.067111  1B
23640     Chris     Gimenez  510000  13  107  0.064806  1B
23356      Kole     Calhoun  506000  90  493  0.064275  1B
23244   Stephen        Vogt  502500  26  269  0.063793  1B
23401    Martin   Maldonado  502000  14  111  0.063745  1B
23617      Rene      Rivera  506400  27  294  0.062997  1B
23479      Matt       Adams  516000  55  527  0.062305  1B
23605     Darin         Ruf  500000  13  102  0.062069  1B
23028     Wilin     Rosario  502000  46  382  0.060733  1B

nameFirst  nameLast  salary    R   AB     value POS
23159      Josh  Harrison  513000   77  520  0.067586  2B
23572     Derek  Dietrich  501500   31  158  0.064998  2B
23050    Marwin  Gonzalez  504500   33  285  0.064776  2B
23333      Josh  Rutledge  501000   44  309  0.064559  2B
23411     Brian    Dozier  540000  112  598  0.063920  2B
23581   Scooter   Gennett  504000   55  440  0.063441  2B
23080       Dee    Gordon  515000   92  609  0.063324  2B
23546        DJ  LeMahieu  501000   59  494  0.062775  2B
22975   Eduardo   Escobar  507500   52  433  0.061977  2B
23308    Marcus    Semien  500000   30  231  0.060079  2B

nameFirst   nameLast  salary   R   AB     value POS
23159      Josh   Harrison  513000  77  520  0.067586  SS
23050    Marwin   Gonzalez  504500  33  285  0.064776  SS
23333      Josh   Rutledge  501000  44  309  0.064559  SS
22975   Eduardo    Escobar  507500  52  433  0.061977  SS
23308    Marcus     Semien  500000  30  231  0.060079  SS
23514     Chris     Owings  500000  34  310  0.060000  SS
23610     Jordy     Mercer  515500  56  506  0.059254  SS
23391   Donovan     Solano  511500  26  310  0.058710  SS
23143      Eric     Sogard  510000  38  291  0.058522  SS
23601     Cesar  Hernandez  500000  13  114  0.058065  SS

nameFirst    nameLast  salary   R   AB     value POS
23132      Josh   Donaldson  500000  93  608  0.068489  3B
23159      Josh    Harrison  513000  77  520  0.067586  3B
23325    Lonnie  Chisenhall  510800  62  478  0.067111  3B
23542     Nolan     Arenado  500000  58  432  0.065665  3B
23303     Conor   Gillaspie  512000  50  464  0.065619  3B
23050    Marwin    Gonzalez  504500  33  285  0.064776  3B
23333      Josh    Rutledge  501000  44  309  0.064559  3B
23546        DJ    LeMahieu  501000  59  494  0.062775  3B
23291     Manny     Machado  519000  38  327  0.062401  3B
22975   Eduardo     Escobar  507500  52  433  0.061977  3B

nameFirst   nameLast  salary   R   AB     value POS
23570     Scott  Van Slyke  507500  32  212  0.076094  RF
23159      Josh   Harrison  513000  77  520  0.067586  RF
23417     Aaron      Hicks  507500  22  186  0.067154  RF
23010   Charlie   Blackmon  501000  82  593  0.066845  RF
23556    Robbie   Grossman  504500  42  360  0.066703  RF
23128    Collin    Cowgill  506000  37  260  0.065197  RF
23356      Kole    Calhoun  506000  90  493  0.064275  RF
23244   Stephen       Vogt  502500  26  269  0.063793  RF
23579   Marcell      Ozuna  505000  72  565  0.062771  RF
23519     David      Lough  510500  31  174  0.060509  RF

nameFirst   nameLast  salary   R   AB     value POS
23570      Scott  Van Slyke  507500  32  212  0.076094  CF
23545      Corey  Dickerson  500000  74  436  0.072803  CF
23580  Christian     Yelich  505000  94  582  0.071733  CF
23458   Starling      Marte  500000  73  495  0.071193  CF
23275       Adam      Eaton  511000  76  486  0.070830  CF
23279      A. J.    Pollock  507000  41  265  0.069654  CF
23417      Aaron      Hicks  507500  22  186  0.067154  CF
23010    Charlie   Blackmon  501000  82  593  0.066845  CF
23635    Brandon      Guyer  501200  37  259  0.066739  CF
23556     Robbie   Grossman  504500  42  360  0.066703  CF

nameFirst   nameLast  salary   R   AB     value POS
23570      Scott  Van Slyke  507500  32  212  0.076094  LF
23545      Corey  Dickerson  500000  74  436  0.072803  LF
23580  Christian     Yelich  505000  94  582  0.071733  LF
23458   Starling      Marte  500000  73  495  0.071193  LF
23279      A. J.    Pollock  507000  41  265  0.069654  LF
23159       Josh   Harrison  513000  77  520  0.067586  LF
23417      Aaron      Hicks  507500  22  186  0.067154  LF
23010    Charlie   Blackmon  501000  82  593  0.066845  LF
23635    Brandon      Guyer  501200  37  259  0.066739  LF
23556     Robbie   Grossman  504500  42  360  0.066703  LF

nameFirst   nameLast  salary   R   AB     value POS
23444     Derek     Norris  505000  46  385  0.071395   C
23001     Devin   Mesoraco  525000  54  384  0.068398   C
23640     Chris    Gimenez  510000  13  107  0.064806   C
23244   Stephen       Vogt  502500  26  269  0.063793   C
23401    Martin  Maldonado  502000  14  111  0.063745   C
23594     J. R.     Murphy  502700   7   81  0.063188   C
23617      Rene     Rivera  506400  27  294  0.062997   C
23589    Josmil      Pinto  500000  25  169  0.062944   C
23515      Evan     Gattis  520250  41  369  0.060876   C
23028     Wilin    Rosario  502000  46  382  0.060733   C

nameFirst  nameLast  salary  R  AB     value POS
23547     Jordan     Lyles  502000  4  48  0.046871   P
22572       Jake   Arrieta  500000  3  39  0.043902   P
23489     Shelby    Miller  521000  6  48  0.042226   P
23607     Gerrit      Cole  512500  3  46  0.040650   P
23258  Henderson   Alvarez  525400  2  57  0.038711   P
22944       Wade     Miley  523500  3  57  0.038204   P
23456       Jeff     Locke  513000  3  45  0.029032   P
23535       Tony  Cingrani  512500  0  21  0.027875   P
23650     Tanner     Roark  506100  1  58  0.026792   P
23228      Lance      Lynn  535000  5  60  0.025881   P

## First base

If we spend \$5m next season, Lonnie Chisenhall is our best value pick for a veteran. He had 478 AB while earning \$510,800.

Scott Van Slyke is also worth speaking with. While having less than half the at bats than Chisenhall, this \$30k lower salary make him a very compelling selection.

## Second base

Josh Harrison is a versitale player. He's showing up or leading in our value rankings for 5 differnt positions. It's a long season.

## Short Stop
While another strong position for Josh Harrison, Marwin Gonzalez could be a great pick here. He's also about \$10k cheaper.

## Third Baseman

We need to target Josh Donaldson here. He's a very valuable veteran that made the most out of his 608 AB last season.

This is a position that Harrison or Chisenhall could also play backup for.

## Right Field

While Van Slyke and Harrison are at the top of this chart, it worth taking a look at Aaron Hicks and Charlie Blackmon. We have similiar values for both players.

Blackmon deffiently had more experience, and was slightly cheaper than Hicks. It might be worth reaching past Hicks on this one. 

## Center Field

Corey Dickerson should be our target here if we've already picked up Van Slyke for another position. One of the higher ratings that we have assigned with a decent sample size for ABs.

## Left Field

Christian Yelich and/or Starling Marte would both make great additions and give us a very valuable outfield.

## Catcher

Derek Norris was about 3 points higher than the next player when it came to catcher value. He's also the most experienced of the group.

## Pitcher

Really think we need to look at other stats for pitchers. Can we switch to the AL like the Astros?